### PR TITLE
Support retry logic for auto recovery

### PR DESCRIPTION
**Pull request by wolfstudy**
Descriptions of the changes in this PR:

### Motivation

In the current implementation, when the Session between Bookie and ZK expires, Auto Recovery does not support automatic reconnection. However, the bookie server itself supports reconnection logic, so when Auto Recovery and Bookie Server are deployed together, once the Session expires with ZK, Auto Recovery will automatically stop and cause the entire Bookie service to go down.

However, in a production environment, session expiration between Bookie and ZK itself is a high-frequency operation, such as the following scenario:

From the ZK side, we can see that the current ZK zxid has overflowed, triggering XidRolloverException:

image
image

When the above situation occurs, we can see that the Session between Bookie and ZK expires:

image
image

Since Auto Recovery has no reconnection logic, the entire Bookie service is down.

### Changes

- Add retry logic for auto recovery

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorElector.java [hunk 1]
```diff
@@ -191,8 +192,14 @@ public void run() {
                         Thread.currentThread().interrupt();
                         submitShutdownTask();
                     } catch (Exception e) {
-                        LOG.error("Exception while performing auditor election", e);
-                        submitShutdownTask();
+                        if (e.getCause() instanceof KeeperException.ConnectionLossException
```
Comment (zymap) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorElector.java:195]: It's better to retry with a delay time I think.

Comment (dlg99) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorElector.java:195]: IIRC ConnectionLoss is a non-recoverable exception.
To handle it properly, one has to re-create ZK client and deal with the watchers/cached state. Ephemeral nodes expire as result of such error too.
Re-initializing the client at random place when error happens is a rather big chunk of work IIRC and it easier to bounce the service in this case, hence https://github.com/apache/bookkeeper/pull/3374

**Review by dlg99 — CHANGES_REQUESTED** [review]
1. Please see https://github.com/apache/bookkeeper/pull/3374 and https://cwiki.apache.org/confluence/display/ZOOKEEPER/ErrorHandling
2. IIRC ConnectionLoss (or SessionExpiredEx or both) is a non-recoverable exception. To handle it properly, one has to re-create ZK client and deal with the watchers/cached state. Ephemeral nodes expire as result of such error too. Re-initializing the client at random place when error happens is a rather big chunk of work IIRC and it easier to bounce the service in this case, hence https://github.com/apache/bookkeeper/pull/3374
3. Change without unit tests/repro

If you disagree with my assessment please start a discussion in the mailing list.

@eolivelli has deeper knowledge of zookeeper, let's double-check with him too.

It is not a good idea in general to run the autorecovery as part of the bookie in prod. The best practice is to run it as a separate service on either another host (which allows one to have fewer AR service nodes than bookies) or alongside with the bookie as a separate process. Running AR separately also helps with such issues like AR activity resulting in a more frequent java GC cycles. HTH.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorElector.java (hunk 1 above)
Comment (zymap) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorElector.java:195]: @dlg99 I saw it said connection loss is a recoverable exception.
```
Recoverable errors: the disconnected event, connection timed out, and the connection loss exception are examples of recoverable errors
```

And the code
```
       public static boolean isRecoverableException(int rc) {
            return KeeperException.Code.CONNECTIONLOSS.intValue() == rc
                    || KeeperException.Code.OPERATIONTIMEOUT.intValue() == rc
                    || KeeperException.Code.SESSIONMOVED.intValue() == rc
                    || KeeperException.Code.SESSIONEXPIRED.intValue() == rc;
        }
```

**Pull request closed without merging**